source/slang/slang-emit-wgsl.cpp

```diff
@@ -0,0 +1,1013 @@
+#include "slang-emit-wgsl.h"
+
+// A note on row/column "terminology reversal".
+//
+// This is an "terminology reversing" implementation in the sense that
+// * "column" in Slang code maps to "row" in the generated WGSL code, and
+// * "row" in Slang code maps to "column" in the generated WGSL code.
+//
+// This means that matrices in Slang code end up getting translated to
+// matrices that actually represent the transpose of what the Slang matrix
+// represented.
+// Both API's adopt the standard matrix multiplication convention whereby the
+// column count of the matrix on the left hand side needs to match row count of
+// the matrix on the right hand side.
+// For these reasons, and due to the fact that (M_1 ... M_n)^T = M_n^T ... M_1^T,
+// the order of matrix (and vector-matrix products) products must also reversed
+// in the WGSL code.
+//
+// This may lead to confusion (which is why this note is referenced in several
+// places), but the benefit of doing this is that the generated WGSL code is
+// simpler to generate and should be faster to compile.
+// A "terminology preserving" implementation would have to generate lots of
+// 'transpose' calls, or else perform more complicated transformations that
+// end up duplicating expressions many times.
+
+namespace Slang {
+
+void WGSLSourceEmitter::emitSwitchCaseSelectorsImpl(
```

Comment:
We shouldn't be handling this logic here. The frontend should insert the necessary coercion logic to make sure case values have the same type as their condition.

Reply:
You just need to modify `SemanticsStmtVisitor::visitCaseStmt` into:
```
    void SemanticsStmtVisitor::visitCaseStmt(CaseStmt* stmt)
    {
        auto expr = CheckExpr(stmt->expr);
        auto switchStmt = FindOuterStmt<SwitchStmt>();

        if (!switchStmt)
        {
            getSink()->diagnose(stmt, Diagnostics::caseOutsideSwitch);
        }
        else
        {
            // coerce to type being switch on, and ensure that value is a compile-time constant
            // The Vals in the AST are pointer-unique, making them easy to check for duplicates
            // by addeing them to a HashSet.
            expr = coerce(CoercionSite::General, switchStmt->condition->type, expr);
        }

        auto exprVal = tryConstantFoldExpr(expr, ConstantFoldingKind::CompileTime, nullptr);
        stmt->expr = expr;
        stmt->exprVal = exprVal;
        stmt->parentStmt = switchStmt;
    }
```

Reply:
(Done.)

This is #4921, which I hoped to defer because I hadn't thought through whether that would break something.

The de-facto behavior of Slang is currently that of GLSL, which will coerce signed to unsigned in case of mismatch, regardless of which is the condition and which is the case expression.

However I don't immediately see how just trying to coerce the case to be of the same type as the condition could cause semantics differing from the current behavior, so I removed this back-end coercion and added a patch to do it in the front-end.